### PR TITLE
API-18995/  loast  add spectral rules and cleanup results

### PR DESCRIPTION
**Pull request by mheusgit**
- Implemented core-ruleset support for Spectral in LOAST.  Some rules were either elevated in severity or disabled
- Added 11 custom rules
- Renamed ruleset file
- Added ability to get rawSchema from OASSchema since Swagger would make small enough changes to schema object to break Spectral checks
- Expanded on sanitize logic to deal with as many scenarios as possible from Spectral

### CODE REVIEW

**Review by dsteeve — APPROVED** [review]
looks good!

**Review by christineyost — COMMENTED** [review]
Ran this on the facilities OAS and saw a lot of duplication of errors for schemas that are present in components -> schemas. For example, if an `Hours` schema is used for all operations and is defined in components, it looks like we'd get the same error for every operation as well as for the component itself. Is it possible to avoid this duplication?

**Comment by mheusgit** [comment]
> Ran this on the facilities OAS and saw a lot of duplication of errors for schemas that are present in components -> schemas. For example, if an `Hours` schema is used for all operations and is defined in components, it looks like we'd get the same error for every operation as well as for the component itself. Is it possible to avoid this duplication?

Have a look at the "Observations" comment in API-18995

File: src/suites/oas-ruleset/validation/oas-ruleset-validator.ts [hunk 1]
```diff
@@ -33,17 +95,8 @@ class OasRulesetValidator extends BaseValidator {
...
-    }
-
-    const ruleMap = this.operationMap.get(operation);
-
-    if (!ruleMap?.has(ruleName)) {
-      ruleMap?.set(ruleName, { failures: new Map(), warnings: new Map() });
-    }
-
     const message = new OasRulesetMessage(type, path, props);
+    const ruleMap = this.registerRule(operation, ruleName);
```
[the remark this answers is not in the document]
Comment (mheusgit) [reply src/suites/oas-ruleset/validation/oas-ruleset-validator.ts:99]: New rules will be registered here but not often.  When they are they would be part of the 'high level' problems existing outside of Spectral's core ruleset.  These problems tend to be for issues that prevent Spectral from being able to figure out what ruleset it should use.

[the remark this answers is not in the document]
Comment (mheusgit) [reply src/suites/oas-ruleset/validation/oas-ruleset-validator.ts:99]: The registerRule() acts as a fallback in these uncommon cases but normally it just getting the ruleMap

File: src/suites/oas-ruleset/validation/oas-ruleset-validator.ts [hunk 2]
```diff
@@ -84,27 +189,88 @@ class OasRulesetValidator extends BaseValidator {
...
+        result.path as string[],
+      );
 
-      // Get operation string (example: /path:GET)
-      let operation = newPath[0];
-      if (newPath.length > 2) {
-        operation = newPath[0] + ':' + newPath[1].toUpperCase();
+      if (operation === operationEnum.openapidoc) {
+        // Indicates some manner of high level validation error
+        msgType = Type.OasRulesetError;
```
[the remark this answers is not in the document]
Comment (mheusgit) [reply src/suites/oas-ruleset/validation/oas-ruleset-validator.ts:198]: Yes.  One scenario I ran into was getting a Spectral warning back about bad openapidoc versions being passed.  The error ID was not part of the core-ruleset so I couldn't override it in the ruleset.yaml

The logic is forcing that to be flagged as a error.

File: src/suites/oas-ruleset/validation/oas-ruleset-validator.ts [hunk 3]
```diff
@@ -84,27 +189,88 @@ class OasRulesetValidator extends BaseValidator {
...
 
-    return path;
+    if (path.length < 2) {
+      // Root level scenario: info, tags, openapi, or paths
+      details = {
+        operation: `ROOT:${path[0]}`,
+        cleanedPath: [] as string[],
+      };
+    } else {
+      let operation = path[0].toUpperCase() + ':' + path[1];
```
[the remark this answers is not in the document]
Comment (mheusgit) [reply src/suites/oas-ruleset/validation/oas-ruleset-validator.ts:235]: That is the beginning of the scenarios that are not root level.  So the problem is deeper inside info, tags, paths, servers, components and not just being reported at the top.

**Comment by mheusgit** [comment]
I've added media type grouping to reduce the number of redundant items reported for endpoints..

**Comment by mheusgit** [comment]
Spectral does not return the original reference it used to resolve paths.  

For example the original errors path is: WaitTimes -> example
But the following is referenced similar
Properties -> properties -> wait_times -> example
GeoFacility -> properties -> properties -> properties -> wait_times -> example

I don't believe there is a way to combine the paths without writing custom function overrides of Spectral.

**Comment by mheusgit** [comment]
PR should be good to go

File: test/suites/oas-ruleset/validation/oas-ruleset-validator.test.ts [hunk 4]
```diff
@@ -103,7 +139,7 @@ describe('OasRulesetValidator', () => {
       const validator = new OasRulesetValidator(oasSchema);
       await validator.validate();
 
-      expect(validator.operationMap.size).toEqual(3);
+      expect(validator.operationMap.size).toEqual(9);
```
[the remark this answers is not in the document]
Comment (dsteeve) [reply test/suites/oas-ruleset/validation/oas-ruleset-validator.test.ts:142]: agree with this comment.  The goal of this test seems to be checking the sanitization logic, but we aren't really doing that? or is that somewhere else?

File: test/suites/oas-ruleset/validation/oas-ruleset-validator.test.ts [hunk 5]
```diff
@@ -103,7 +139,7 @@ describe('OasRulesetValidator', () => {
       const validator = new OasRulesetValidator(oasSchema);
```
Comment (dsteeve) [inline test/suites/oas-ruleset/validation/oas-ruleset-validator.test.ts:139]: on the line above, it should be not "three" but "nine"?

**Review by stevelong00 — COMMENTED** [review]
Looks pretty good!

**Review by christineyost — APPROVED** [review]
Thanks for updating the test!

**Pull request merged**